src/snowflake/snowpark/modin/plugin/_internal/snowpark_pandas_types.py

```diff
@@ -0,0 +1,141 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import datetime
+import inspect
+from abc import ABCMeta, abstractmethod
+from typing import Any, Callable, Optional, Union
+
+import numpy as np
+import pandas as native_pd
+
+from snowflake.snowpark.modin.plugin.utils.warning_message import WarningMessage
+from snowflake.snowpark.types import DataType, LongType
+
+TIMEDELTA_WARNING_MESSAGE = (
+    "Snowpark pandas support for Timedelta is not currently available."
+)
+
+
+_python_type_to_from_pandas: dict[type, Callable[[Any], Any]] = {}
+_pandas_type_to_snowpark_pandas_type: dict[type, "SnowparkPandasType"] = {}
+
+
+class SnowparkPandasTypeMetaclass(
+    # Inherit from ABCMeta so that this metaclass makes ABCs.
+    ABCMeta
+):
+    """
+    This class is a Metaclass for Snowpark pandas types.
+
+    Defining a class through this metaclass updates some global type conversion
+    information. We can refer to that information anywhere we need to do Snowpark
+    pandas type conversion, e.g. in from_pandas and to_pandas.
+    """
+
+    def __new__(cls: type, clsname: str, bases: Any, attrs: Any) -> type:
+        # Create a type using this class's superclass. mypy raises the error
+        # 'Argument 2 for "super" not an instance of argument 1', which is
+        # difficult to fix.
+        new_snowpark_python_type = super().__new__(cls, clsname, bases, attrs)  # type: ignore
+
+        if inspect.isabstract(new_snowpark_python_type):
+            return new_snowpark_python_type
+
+        for cls in new_snowpark_python_type.types_to_convert_with_from_pandas:
+            for existing_cls in _python_type_to_from_pandas:
+                # we don't want any class in _python_type_to_from_pandas to be
+                # a subclass of another type in _python_type_to_from_pandas.
+                # Otherwise, the rewriting rules for the two types may
+                # conflict.
+                assert not issubclass(
+                    cls, existing_cls
+                ), f"Already registered from_pandas for class {cls} with {existing_cls}"
+            _python_type_to_from_pandas[cls] = new_snowpark_python_type.from_pandas
+
+        for existing_cls in _pandas_type_to_snowpark_pandas_type:
+            # we don't want any class in _pandas_type_to_snowpark_pandas_type to be
+            # a subclass of another type in _pandas_type_to_snowpark_pandas_type.
+            # Otherwise, the conversions rules for the two types may conflict.
+            assert not issubclass(
+                new_snowpark_python_type.pandas_type, existing_cls
+            ), f"Already registered Snowpark pandas type for class {cls} with {existing_cls}"
+        _pandas_type_to_snowpark_pandas_type[
+            new_snowpark_python_type.pandas_type
+        ] = new_snowpark_python_type
+
+        return new_snowpark_python_type
+
+
+class SnowparkPandasType(DataType, metaclass=SnowparkPandasTypeMetaclass):
```

Comment:
Sorry I missed this earlier. This class also needs to be frozen dataclass (for immutability) as this is being used inside InternalFrame. Please take a look at OrderingColumn class for example.

Reply:
Fixed.